Content Helper: Add Excerpt Generator feature

vaurdan opened the pull request:
<!--
Hello! Just a quick reminder that this is a public repo. Please don't include any internal links or sensitive data (like PII, private code, client names, site URLs, etc. If you're not sure if something is safe to share, please just ask!
-->

## Description
This pull request introduces the Excerpt Generator feature within the Gutenberg editor, powered by Parse.ly AI. This tool provides editors with automated, AI-driven excerpt suggestions to enhance their content creation process. 

Additionally, this PR brings a new `BetaBadge` component, which is utilized to label the Excerpt Generator feature as 'Beta' in the user interface. This badge aims to set the correct expectations for users and invites them to provide feedback on this new functionality. The `BetaBadge` is flexible and can be customized with different text, colors, and sizes, making it a reusable component for future beta features, such as the Title Suggestions feature.

While integrating this feature, particular attention has been paid to ensure compatibility with other plugins offering similar functionality, most notably Jetpack. The current implementation gracefully deactivates Jetpack's excerpt feature to prevent overlap, with ongoing considerations about how best to give users control over which excerpt tool they prefer to use.


## Motivation and context
The Excerpt Generator aims to optimize the editorial workflow, providing quick and intelligent excerpt suggestions that can improve SEO and audience engagement. The inclusion of the `BetaBadge` sets clear expectations regarding the feature's developmental stage and encourages user involvement in its evolution.

## How has this been tested?
The new features were tested in a local WordPress environment running the latest Gutenberg editor. 

- Functionality testing confirmed that the Excerpt Generator performs as expected within the Gutenberg editor.
- UI consistency checks ensured the `BetaBadge` is displayed correctly.

Still to test:
- [x] Compatibility tests with Jetpack ensured that when both plugins are active, there is no conflict, and the Parse.ly feature operates smoothly.

Edit 12 Dec: I just tested this branch with a site that has Jetpack AI, and it's working nicely:
 - By default, the Jetpack excerpt panel is disabled, in favor of Parse.ly's panel
 - If the `wp_parsely_enable_content_helper_excerpt_generator` is set to false, Jetpack excerpt panel is rendered instead.

## Screenshots (if appropriate)
Here's a screencast of the feature at the current stage:

https://github.com/Parsely/wp-parsely/assets/5804923/0b6ae7cc-e927-458d-a155-b4235fa1a3e1

vaurdan committed:
Add excerpt generator feature to the block editor

acicovic committed:
Merge branch 'develop' into add/feature-excerpt-generator

acicovic commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -130,25 +130,15 @@ protected function register_endpoint( string $endpoint ): void {
-	 * @param string          $param_item         The param element to use to
-	 *                                            get the items.
+	 * @param string|null     $param_item         The param element to use to
+	 *                                       get the items.
```
```suggestion
	 * @param string|null     $param_item         The param element to use to get the items.
```

acicovic commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -130,25 +130,15 @@ protected function register_endpoint( string $endpoint ): void {
 	 * @param WP_REST_Request $request            The request object.
 	 * @param bool            $require_api_secret Specifies if the API Secret is
 	 *                                            required.
```
```suggestion
	 * @param bool            $require_api_secret Specifies if the API Secret is required.
```

acicovic commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -173,6 +163,35 @@ protected function get_data( WP_REST_Request $request, bool $require_api_secret
 	}
 
+	/**
+	 * Validates that the API key and secret are set.
```
Let's use the `Site ID` terminology (instead of `API key`) here and anywhere else applicable.

acicovic commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -173,6 +163,35 @@ protected function get_data( WP_REST_Request $request, bool $require_api_secret
+	 *
+	 * @return WP_Error|bool
+	 */
+	protected function validate_apikey_and_secret( bool $require_api_secret = true ) {
```
Nice move moving this into its own function!

acicovic commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -173,6 +163,35 @@ protected function get_data( WP_REST_Request $request, bool $require_api_secret
+				array( 'status' => 403 )
+			);
+		}
+		return true;
```
Consider adding a blank line above the `return`.

acicovic commented on wp-parsely.php:
```diff
@@ -160,6 +163,7 @@ function parsely_wp_admin_early_register(): void {
 require_once __DIR__ . '/src/Endpoints/class-related-api-proxy.php';
 require_once __DIR__ . '/src/Endpoints/class-rest-metadata.php';
 require_once __DIR__ . '/src/Endpoints/content-suggestions/class-write-title-api-proxy.php';
+require_once __DIR__ . '/src/Endpoints/content-suggestions/class-suggest-meta-description-api-proxy.php';
```
Let's add this above `class-write-title-api-proxy.php`, so it's sorted alphabetically. Same for line 183.

acicovic commented on wp-parsely.php:
```diff
@@ -244,6 +255,17 @@ function init_content_helper_editor_sidebar(): void {
 
+require_once __DIR__ . '/src/content-helper/excerpt-generator/class-excerpt-generator.php';
+/**
+ * Initialize and inserts the PCH Excerpt Generator.
```
```suggestion
 * Initializes and inserts the PCH Excerpt Generator.
```

acicovic commented on wp-parsely.php:
```diff
@@ -244,6 +255,17 @@ function init_content_helper_editor_sidebar(): void {
+function init_content_helper_excerpt_generator(): void {
+	( new Excerpt_Generator( $GLOBALS['parsely'] ) )->run();
+}
+add_action( 'init', __NAMESPACE__ . '\\init_content_helper_excerpt_generator' );
```
Let's move this `add_action` above `init_content_helper_excerpt_generator()` to keep the file's convention, unless there's an advantage doing it this way.

acicovic commented on src/RemoteAPI/content-suggestions/class-content-suggestions-base-api.php:
```diff
@@ -36,4 +39,53 @@ protected function get_request_options(): array {
+	 *
+	 * @param array<string|int|bool> $query An associative array containing the query parameters for the API request.
+	 * @param array<string|int|bool> $body An associative array containing the body parameters for the API request.
+	 *
```
```suggestion
```

acicovic commented on src/RemoteAPI/content-suggestions/class-content-suggestions-base-api.php:
```diff
@@ -36,4 +39,53 @@ protected function get_request_options(): array {
+	 * This method sends a POST request to the Parse.ly Content Suggestion API and returns the response.
+	 * The response is either a WP_Error object in case of an error, or a decoded JSON object in case of a
+	 * successful request.
+	 *
```
```suggestion
	 *
	 * @since 3.13.0
	 *
```

acicovic commented on src/RemoteAPI/content-suggestions/class-suggest-meta-description-api.php:
```diff
@@ -0,0 +1,63 @@
+ * @package Parsely
+ * @since 3.13.0
+ */
+
```
```suggestion

declare(strict_types=1);

```

acicovic commented on src/RemoteAPI/content-suggestions/class-suggest-meta-description-api.php:
```diff
@@ -0,0 +1,63 @@
+<?php
+/**
+ * Remote API: Content Suggestions Suggest Meta Description  API
```
```suggestion
 * Remote API: Content Suggestions Suggest Meta Description API
```

acicovic commented on src/RemoteAPI/content-suggestions/class-suggest-meta-description-api.php:
```diff
@@ -0,0 +1,63 @@
+	 *
+	 * @param string $title   The title of the content.
+	 * @param string $content The query arguments to send to the remote API.
+	 *
```
```suggestion
```

No need for blank line between `@param` and `@return` in PHP files specifically.

acicovic commented on src/content-helper/excerpt-generator/provider.ts:
```diff
@@ -0,0 +1,40 @@
+ */
+import { ContentHelperError, ContentHelperErrorCode } from '../common/content-helper-error';
+
+interface ExcerptGeneratorApiResponse {
```
Consider adding a DocBlock here, as well as for the class below. You can look at `PerformanceDetailsProvider` for an example. I'm aware that other providers miss similar comments.

acicovic commented on src/content-helper/common/css/variables.scss:
```diff
@@ -46,4 +47,5 @@ $html-font-size: 16px; // Used in functions.scss.
 	--font-size--extra-large: 1.2rem;
 	--black: #000;
 	--sidebar-black: #1e1e1e;
+	--parsely-green: #5ba745;
```
Does this map to the `--parsely-green: hsl(107, 42%, 46%);` present in the original Parse.ly stylesheet?

vaurdan replied:
Actually no, I haven't seen the original Parse.ly stylesheet. However, I got the color from the parse.ly SVG logo. Seems to be the same color, but because of consistency I will replace it with the hsl color.

acicovic commented on src/content-helper/common/components/beta-badge.tsx:
```diff
@@ -0,0 +1,25 @@
+
+export const BetaBadge = ( {
+	text = 'Beta',
+	color = 'var(--parsely-green)', // Default orange color
```
`orange` seems incorrect here. We can correct this to `green` or remove the word altogether. We should also put periods after all comments in this file.

acicovic commented on src/content-helper/common/components/beta-badge.tsx:
```diff
@@ -0,0 +1,25 @@
+	color?: string;
+	size?: string;
+};
+
```
A short DockBlock describing what the BetaBadge is could be useful.

acicovic commented on src/content-helper/common/icons/leaf-icon.tsx:
```diff
@@ -3,11 +3,11 @@
 	<SVG
-		height={ 24 }
+		height={ size }
 		viewBox="0 0 60 65"
```
Have we checked whether the `viewBox` needs to be changed, relative to the new `size` variable? If not, let's verify this.

vaurdan replied:
I believe the viewBox attribute should remain unchanged even if the size is adjusted. From what I understand, the viewBox maintains the SVG's aspect ratio and coordinate system, allowing it to scale correctly.

acicovic commented on src/content-helper/excerpt-generator/excerpt-generator.scss:
```diff
@@ -0,0 +1,55 @@
+		margin-bottom: to_rem(10px);
+
+		.wp-parsely-excerpt-generator-header-label {
+			font-size: 11px;
```
```suggestion
			font-size: to_rem(11px);
```

acicovic commented on src/@types/global.d.ts:
```diff
@@ -0,0 +1,17 @@
+declare global {
```
I think it would be nice to have a short comment on why we're implementing this, so someone without prior knowledge can get an idea.

acicovic replied:
Sorry, I didn't explain what I meant. I was referring to commenting about why we need to track Jetpack Editor initial state. If you feel that this would be useful, add it.

acicovic commented on src/content-helper/excerpt-generator/components/excerpt-panel.tsx:
```diff
@@ -0,0 +1,197 @@
+/**
+ * External dependencies
```
```suggestion
 * WordPress dependencies
```

Conventionally, we've been using `WordPress dependencies` when imports are only coming from `@wordpress` packages. There are a couple of additional occurrences of this.

acicovic commented on src/content-helper/excerpt-generator/components/excerpt-panel.tsx:
```diff
@@ -0,0 +1,197 @@
+import { ContentHelperError } from '../../common/content-helper-error';
+import { BetaBadge } from '../../common/components/beta-badge';
+
+const PostExcerptGenerator = () => {
```
Consider adding class descriptions and `@since` DocBlocks where applicable.

acicovic commented on src/content-helper/excerpt-generator/components/excerpt-panel.tsx:
```diff
@@ -0,0 +1,197 @@
+								variant="secondary"
+								onClick={ acceptGeneratedExcerpt }
+							>
+								Accept
```
Let's use `__()` on the `Accept` and `Discard` strings.

acicovic commented on src/content-helper/excerpt-generator/excerpt-generator.tsx:
```diff
@@ -0,0 +1,44 @@
+ */
+import { ExcerptPanel } from './components/excerpt-panel';
+
+// TODO: Get the plugin ID from the editor sidebar file.
```
Can we to act on this `TODO` in this PR, or is this for another time?

vaurdan replied:
Addressing this TODO in this PR would require exporting `BLOCK_PLUGIN_ID` from the `src/content-helper/editor-sidebar/editor-sidebar.tsx` file and importing in the `ExcerptGenerator` component. However, this would result in the entire editor sidebar code being included in the excerpt generator JS bundle, leading to duplicate loading. To avoid this, we would need to refactor our script packaging and enqueueing process. Given the scope of these changes, it would be more appropriate to handle this in a separate PR.

acicovic replied:
Makes sense. Could you post this as an issue so we don't lose track of it, or note it to make a PR for this in the near future?

acicovic reviewed (commented):
Great work as always!

I've done my first pass of the code and I'll mostly bother you with details as always! 🙂

When you look into addressing the tests issue, also please look into whether any SonarCloud warnings need resolution. 🙂

vaurdan committed:
Fix tests

vaurdan committed:
Add integration test for `Suggest_Meta_Description_API` class

vaurdan committed:
Merge branch 'develop' into add/feature-excerpt-generator

# Conflicts:
#       package-lock.json
#       package.json

vaurdan committed:
Fix linting

vaurdan committed:
Apply suggestions from code review

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

vaurdan committed:
Apply remaining code suggestion fixes

vaurdan committed:
Merge branch 'develop' into add/feature-excerpt-generator

acicovic committed:
Merge branch 'develop' into add/feature-excerpt-generator

acicovic commented on src/Endpoints/class-base-api-proxy.php:
```diff
@@ -173,6 +161,36 @@ protected function get_data( WP_REST_Request $request, bool $require_api_secret
+	 * @since 3.13.0
+	 *
+	 * @param bool $require_api_secret Specifies if the API Secret is required.
+	 *
```
```suggestion
```

acicovic commented on src/RemoteAPI/content-suggestions/class-suggest-meta-description-api.php:
```diff
@@ -0,0 +1,64 @@
+	/**
+	 * Indicates whether the endpoint is public or protected behind permissions.
+	 *
+	 * @since 3.12.0
```
```suggestion
	 * @since 3.13.0
```

acicovic commented on src/content-helper/common/components/beta-badge.tsx:
```diff
@@ -0,0 +1,44 @@
+ * The BetaBadge component displays a badge with the specified text, color, and size.
+ *
+ * @param {BetaBadgeProps} props - The properties for the BetaBadge component.
+ * @return {JSX.Element} The BetaBadge component.
```
```suggestion

 * @return {JSX.Element} The BetaBadge component.
```

The tricky part here is that for JS we have a blank line before `@param` and `@return`, and for PHP we don't. This is a detail to remember, as unfortunately this is how things are in WP inline documentation standards. Otherwise, we can discuss having a common approach for everything.

acicovic commented on src/content-helper/common/components/beta-badge.tsx:
```diff
@@ -0,0 +1,44 @@
+ * @return {JSX.Element} The BetaBadge component.
+ */
+export const BetaBadge = ( {
+	text = 'Beta',
```
Let's use `__()` to localize this, just in case.

acicovic commented on src/content-helper/excerpt-generator/components/excerpt-panel.tsx:
```diff
@@ -0,0 +1,214 @@
+
+/**
+ * The ExcerptPanel component verifies that the current post type supports excerpts,
+ * and then renders the PostExcerptGenerator component.
```
```suggestion
 * and then renders the PostExcerptGenerator component.
 *
 * @since 3.13.0
```

acicovic commented on src/content-helper/excerpt-generator/excerpt-generator.tsx:
```diff
@@ -0,0 +1,44 @@
+// TODO: Get the plugin ID from the editor sidebar file.
+const PARSELY_SIDEBAR_PLUGIN_ID = 'wp-parsely-block-editor-sidebar';
+
+const ExcerptGenerator = ( settings: never, name: string ) => {
```
Let's add a DocBlock here, including a `@since`.

acicovic commented on src/content-helper/excerpt-generator/provider.ts:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Specifies the form of the response returned by the
+ * `/content-suggestions/suggest-meta-description` WordPress REST API endpoint.
+ */
```
```suggestion
 *
 * @since 3.13.0
 */
```

acicovic commented on src/content-helper/excerpt-generator/provider.ts:
```diff
@@ -0,0 +1,55 @@
+
+/**
+ * Provides the generate excerpt functionality to be used in other components.
+ */
```
```suggestion
 *
 * @since 3.13.0
 */
```

acicovic commented on wp-parsely.php:
```diff
@@ -176,6 +180,7 @@ function parsely_wp_admin_early_register(): void {
 require_once __DIR__ . '/src/RemoteAPI/class-related-api.php';
 require_once __DIR__ . '/src/RemoteAPI/class-validate-api.php';
 require_once __DIR__ . '/src/RemoteAPI/content-suggestions/class-write-title-api.php';
+require_once __DIR__ . '/src/RemoteAPI/content-suggestions/class-suggest-meta-description-api.php';
```
```suggestion
require_once __DIR__ . '/src/RemoteAPI/content-suggestions/class-suggest-meta-description-api.php';
require_once __DIR__ . '/src/RemoteAPI/content-suggestions/class-write-title-api.php';
```

acicovic commented on tests/Integration/RemoteAPI/content-suggestions/WriteTitleAPITest.php:
```diff
@@ -11,6 +11,7 @@
 namespace Parsely\Tests\Integration\RemoteAPI\ContentSuggestions;
 
 use Parsely\Parsely;
+use Parsely\RemoteAPI\ContentSuggestions\Suggest_Meta_Description_API;
```
```suggestion
```
This doesn't seem to be used by anything?

acicovic commented on src/content-helper/common/css/variables.scss:
```diff
@@ -46,4 +47,5 @@ $html-font-size: 16px; // Used in functions.scss.
 	--font-size--extra-large: 1.2rem;
 	--black: #000;
 	--sidebar-black: #1e1e1e;
+	--parsely-green: hsl(107, 42%, 46%);
```
Great! Let's move this below line 18. This way, we'll exactly match the original stylesheet's order.

acicovic reviewed (commented):
I've made my second pass of the code and left some comments.

acicovic committed:
Merge branch 'develop' into add/feature-excerpt-generator

vaurdan committed:
Apply suggestions from code review

Co-authored-by: Alex Cicovic <23142906+acicovic@users.noreply.github.com>

vaurdan committed:
Apply remaining code suggestion fixes

vaurdan committed:
Add `wp_parsely_use_default_excerpt` filter to allow overriding/disabling the Parse.ly's excerpt implementation

vaurdan commented on src/@types/assets/window.d.ts:
```diff
@@ -27,5 +27,25 @@ declare global {
+		 *
+		 * @see https://github.com/Automattic/jetpack/blob/4eb6a42833879b30aa2a7f4c82e44fc094307de3/projects/plugins/jetpack/extensions/plugins/ai-content-lens/editor.js#L16
+		 */
+		Jetpack_Editor_Initial_State?: {
```
I just realised that we already have a `window.d.ts` file for declaring `window` types, so I ended up moving the `Jetpack_Editor_Initial_State` here, but also introduced a comment further explaining it.

acicovic replied:
Yes, that's great!

vaurdan commented:
@acicovic I addressed all your feedback (thanks for that!) and also included a `wp_parsely_use_default_excerpt` filter, that should be set to true if the Parse.ly Excerpt implementation is to be overridden, either by the core implementation, or other third-party implementation such as Jetpack. I tried to use a generic name so it can be used in different scenarios, but let me know if you think the filter name should be different.

vaurdan commented:
I just tested this branch with a site that has Jetpack AI, and it's working nicely:
 - By default, the Jetpack excerpt panel is disabled, in favor of Parse.ly's panel
 - If the `wp_parsely_use_default_excerpt` is set to true, Jetpack excerpt panel is rendered instead.

acicovic commented:
> also included a wp_parsely_use_default_excerpt filter

I'm sorry, this was an oversight from my part and I think we might already have something that will work. Can you see if the feature gets disabled by setting the `wp_parsely_enable_content_helper_excerpt_generator` filter to `false`?

acicovic commented:
See [here](https://github.com/Parsely/wp-parsely/blob/5fbb96f9028963e12ec6bfb2a11dd57accc9fe35/src/content-helper/common/class-content-helper-feature.php#L75-L114) and [here](https://github.com/Parsely/wp-parsely/blob/5fbb96f9028963e12ec6bfb2a11dd57accc9fe35/src/content-helper/excerpt-generator/class-excerpt-generator.php#L35-L44) for a glimpse of how that works.

acicovic commented:
I've also given the feature a try and it works well for me. So the last thing remaining is whether we'll be removing the filter or not.

vaurdan commented:
@acicovic Ah good point! I just tested with `wp_parsely_enable_content_helper_excerpt_generator` and it works pretty good. I'm going to remove the filter I added, and we can use this one instead. Thanks for catching that :)

vaurdan committed:
Revert "Add `wp_parsely_use_default_excerpt` filter to allow overriding/disabling the Parse.ly's excerpt implementation"

This reverts commit 5fbb96f9028963e12ec6bfb2a11dd57accc9fe35.

acicovic reviewed (approved):
Excellent work, and thanks for bearing with me!
